### ISSUE TITLE
[RFC] PD stage 1: host pool and out-of-band side channel

### ISSUE THREAD

**Issue by cjx0709** [sgl-project/sglang-jax#1242]
## Summary

Introduce the second PD increment on top of the transfer foundation:

- `QueueHostKVPool` as the host-side staging pool abstraction
- `ZmqPullNotifier` for the D -> P pull-done side channel
- event-driven sender/receiver completion, so transfer lifecycle no longer depends on a synchronous placeholder wait after `register_pull()`
- `RequestTransportCore` extraction — domain-agnostic lifecycle management (registries, reaper, terminal records, graceful shutdown) with zero dependency on JAX, KV, or ZMQ
- multi-entry payload refactor — payload changed from `jax.Array` to `dict[str, jax.Array]` with sub-uuid scheme `f"{base_uuid}:{entry_name}"`, enabling hybrid KV (SWA + recurrent state) and EPD embedding transfer

This stage is about the **buffer / side-channel / transport architecture foundation**. Scheduler plumbing and production defaults are handled later.

## Position in the roadmap

This issue defines the second increment in the PD roadmap:

1. Transfer foundation
2. Host pool + side channel + transport core + multi-entry payload *(this issue)*
3. Scheduler + end-to-end
4. Multi-host + routing
5. Production hardening

It builds directly on Stage 0 and prepares the lifecycle primitives that the scheduler will later consume.

## Proposed structure changes

Add or update the following modules:

- `python/sgl_jax/srt/mem_cache/host_kv_pool.py`
- `python/sgl_jax/srt/disaggregation/jax_transfer/zmq_notifier.py`
- `python/sgl_jax/srt/disaggregation/jax_transfer/conn.py`
- `python/sgl_jax/srt/disaggregation/transport/__init__.py` — public surface for transport layer
- `python/sgl_jax/srt/disaggregation/transport/core.py` — `RequestTransportCore` + `TerminalTransferRecord`
- `python/sgl_jax/srt/disaggregation/transport/backend.py` — `TransferBackend` ABC + `JaxTransferBackend`
- `python/sgl_jax/test/disaggregation/test_queue_host_kv_pool.py`
- `python/sgl_jax/test/disaggregation/test_zmq_pull_notifier.py`
- `python/sgl_jax/test/disaggregation/test_kv_sender_event_driven.py`
- `python/sgl_jax/test/disaggregation/test_orphan_reaper.py`
- `python/sgl_jax/test/disaggregation/test_request_transport_core.py` — 19 tests for core in isolation
- `python/sgl_jax/test/disaggregation/test_transfer_backend.py` — 6 tests for backend ABC

### 1. `QueueHostKVPool`

Provide a short-lived pinned-host staging pool abstraction with:

- fixed-size preallocated buffers
- FIFO-style borrow / return semantics
- a device-to-host staging helper for future path-A integration

The abstraction should remain backend-agnostic enough that later host-cache implementations can share the same contract.

### 2. `ZmqPullNotifier`

Introduce a D -> P side-channel for transfer completion:

- P registers a callback for a transfer identity
- D sends a `pull-done` notification after a successful pull
- P uses that callback to release transfer-local resources safely

### 3. Event-driven transfer lifecycle

Update sender/receiver behavior so lifecycle completion is driven by callbacks and polling rather than a synchronous wait placeholder.

Target sender behavior:

```text
WAITING_FOR_INPUT -> TRANSFERRING -> SUCCESS/FAILED
```

Target receiver behavior:

```text
WAITING_FOR_INPUT -> TRANSFERRING -> SUCCESS/FAILED
```

### 4. `RequestTransportCore` extraction

Extract domain-agnostic lifecycle management from `JaxTransferKVManager` into a standalone `RequestTransportCore` class:

- sender/receiver registries (register, prune, inflight count)
- terminal transfer records (bounded OrderedDict, max 4096)
- reaper thread with configurable ack/pull timeouts
- graceful shutdown with drain timeout
- duck-type participant contract: `transfer_started_at` + `fail(reason=...)`
- zero dependency on JAX, KV, ZMQ, or any payload type

`JaxTransferKVManager` becomes a thin façade composing `RequestTransportCore` + `TransferBackend` + `ZmqPullNotifier`.

### 5. Multi-entry payload

Refactor sender/receiver payload from `jax.Array` to `dict[str, jax.Array]`:

- `PMetadata.spec` → `PMetadata.specs: dict[str, jax.ShapeDtypeStruct]`
- `TransferStatus` gains `sub_uuids: tuple[str, ...]`
- `producer_handoff` iterates over entries, registering each under `f"{uuid}:{entry_name}"`
- Sender fail/ack release loops over all sub_uuids
- Receiver poll pulls all entries, checks `all(r.is_ready() ...)` before SUCCESS
- ZMQ ack remains per-request (base uuid); backend ops use sub-uuids
- Entry names must not contain `":"`

Key invariant: **scheduler still sees one req_id = one sender/receiver**. Multi-entry complexity is internal.

### 6. Timeout and orphan hooks

Introduce the primitives needed for later hardening:

- sender/receiver timeout timestamps
- manager-level orphan scanning hooks
- deterministic terminal cleanup behavior

This stage should stop short of full production policy, but the hooks should exist.

### Unit tests (CPU)

- `test_queue_host_kv_pool.py`
  - borrow / return semantics
  - copy-from-device behavior
  - exhaustion behavior
- `test_zmq_pull_notifier.py`
  - callback registration/unregistration
  - duplicate / malformed / unknown message handling
- `test_kv_sender_event_driven.py`
  - sender completion on notifier ack
  - cleanup ownership / no double cleanup
  - multi-entry payload: N entries register N sub-uuids
  - multi-entry fail: releases all sub-uuids
  - empty payload and colon-in-entry-name rejected
- `test_orphan_reaper.py`
  - timeout-driven failure transitions
  - manager cleanup hooks
- `test_request_transport_core.py`
  - register/prune sender and receiver
  - terminal record CRUD + bounded eviction
  - reaper skips not-started, force-fails expired, respects disabled timeouts
  - graceful shutdown
- `test_transfer_backend.py`
  - ABC cannot be instantiated
  - `JaxTransferBackend` delegates correctly
  - isinstance check

## Acceptance criteria

- pinned-host pool abstraction exists and is unit-tested
- side-channel notifier exists and is unit-tested
- sender/receiver completion is event-driven
- `RequestTransportCore` is extracted and independently tested (19 tests)
- `TransferBackend` ABC is extracted and independently tested (6 tests)
- multi-entry payload works with sub-uuid scheme (4 new tests)
- timeout/orphan hooks exist for later hardening stages
- 117 tests pass, ruff clean
